features/apimgt/org.wso2.carbon.apimgt.admin.feature/src/main/resources/admin/source/src/app/components/GatewayEnvironments/AddEditVhost.jsx

```diff
@@ -0,0 +1,282 @@
+import DialogActions from '@material-ui/core/DialogActions';
+import DialogContentText from '@material-ui/core/DialogContentText';
+
+const useStyles = makeStyles((theme) => ({
+    error: {
+        color: theme.palette.error.dark,
+    },
+    vhostPaper: {
+        padding: theme.spacing(1),
+        marginBottom: theme.spacing(1),
+    },
+}));
+
+function AddEditVhost(props) {
+    const { onVhostChange, initialVhosts } = props;
+    const classes = useStyles();
+
+    const [userVhosts, setUserVhosts] = useState(initialVhosts);
+    const [id, setId] = useState(0);
+    const defaultVhost = {
+        host: '', httpContext: '', httpsPort: 443, httpPort: 80, wssPort: 8099, wsPort: 9099, isNew: true,
+    };
+
+    // change handlers
+    const updateChanges = (key, field, newValue) => {
+        const theVhost = userVhosts.find((vhost) => vhost.key === key);
+        theVhost[field] = newValue;
+        let tempItems = userVhosts.filter((vhost) => vhost.key !== key);
+        tempItems = [...tempItems, theVhost];
+        setUserVhosts(tempItems);
+    };
+    const changeHandler = (field) => ({ target: { name, value } }) => {
+        let theValue = value;
+        if (field === 'httpContext') {
+            // remove slashes in start and end of httpContext
+            theValue = value.replace(/(^\/|\/$)/g, '');
+        }
+        updateChanges(name, field, theValue);
+    };
+
+    const [openRemoveVhost, setOpenRemoveVhost] = useState(false);
+    const [selectedKey, setSelectedKey] = useState(0);
+
+    const handleRemoveVhost = (key) => {
+        const theKey = key || selectedKey;
+        const tempItems = userVhosts.filter((vhost) => vhost.key !== theKey);
+        setUserVhosts(tempItems);
+        setOpenRemoveVhost(false);
+    };
+
+    const handleRemoveVhostConfirm = (key, isNew) => {
+        if (isNew) { // not already existing one
+            handleRemoveVhost(key);
+        } else { // already existing vhost
+            setSelectedKey(key);
+            setOpenRemoveVhost(true);
+        }
+    };
+
+    const handleClose = () => {
+        setOpenRemoveVhost(false);
+    };
+
+    const handleNewVhost = () => {
+        const vhost = defaultVhost;
+        vhost.key = '' + id;
+        setId(id + 1);
+        const tempItems = [...userVhosts, vhost];
+        setUserVhosts(tempItems);
+    };
+
+    useEffect(() => {
+        const nonEmptyItems = [];
+        for (const userVhost of userVhosts) {
+            if (userVhost.host && userVhost.host.trim() !== '') {
+                nonEmptyItems.push(userVhost);
+            }
+        }
+        onVhostChange({ target: { name: 'vhosts', value: nonEmptyItems } });
+    }, [userVhosts]);
+
+    useEffect(() => {
+        if (initialVhosts && initialVhosts.length > 0) {
+            let i = 0;
+            setUserVhosts(initialVhosts.map((vhost) => {
+                const keyedVhost = vhost;
+                keyedVhost.key = '' + i++;
```

Comment:
Can't we use vhost key as it's ID , why we need to add index here ? this key = to array index  right ?

Reply:
There is no ID for vhost. Here Host name of VHost is unique, but we are editing it. Used this key to identify Vhosts uniquely, if we use array index then if we push an edited vhost at the end of the array, Vhosts are exchanging its keys (i.e. they are mixing their keys).